File: src/native/dsp/YinPitchDetector.cpp
#include "YinPitchDetector.h"

#include <algorithm>
#include <array>
#include <cmath>
#include <limits>
#include <sstream>

namespace tine::dsp {

namespace {
constexpr double kMinThreshold = 0.001;
constexpr double kMaxThreshold = 0.999;
}

YinPitchDetector::YinPitchDetector(double sampleRate, std::size_t bufferSize, double threshold)
    : m_sampleRate(sampleRate),
      m_bufferSize(bufferSize),
      m_maxLag(bufferSize / 2),
      m_threshold(std::clamp(threshold, kMinThreshold, kMaxThreshold)),
      m_difference(m_maxLag + 1, 0.0),
      m_cumulative(m_maxLag + 1, 1.0) {
    m_lastResult = {};
}

void YinPitchDetector::setThreshold(double threshold) noexcept {
    m_threshold = std::clamp(threshold, kMinThreshold, kMaxThreshold);
}

PitchResult YinPitchDetector::processBuffer(const float* samples, std::size_t numSamples) {
    PitchResult result;

    if (samples == nullptr || numSamples < m_bufferSize || m_maxLag < 2 || m_sampleRate <= 0.0) {
        m_lastResult = result;
        return m_lastResult;
    }

    computeDifference(samples);
    computeCumulativeMeanNormalized();

    double probability = 0.0;
    std::size_t tau = absoluteThreshold(probability);

    if (tau == 0) {
        m_lastResult = result;
        return m_lastResult;
    }

    double refinedTau = static_cast<double>(tau);
    if (tau > 1 && tau < m_maxLag) {
        refinedTau = parabolicInterpolation(tau, m_cumulative);
    }

    if (refinedTau <= 0.0) {
        m_lastResult = result;
        return m_lastResult;
    }

    const double frequency = m_sampleRate / refinedTau;
    if (!std::isfinite(frequency) || frequency <= 0.0) {
        m_lastResult = result;
        return m_lastResult;
    }

    const double midi = midiFromFrequency(frequency);
    const double nearestMidi = std::round(midi);
    const double cents = (midi - nearestMidi) * 100.0;

    result.isValid = probability > 0.0;
    result.frequency = frequency;
    result.midi = midi;
    result.cents = cents;
    result.noteName = noteNameFromMidi(nearestMidi);
    result.probability = std::clamp(probability, 0.0, 1.0);

    m_lastResult = result;
    return m_lastResult;
}

void YinPitchDetector::computeDifference(const float* samples) {
    std::fill(m_difference.begin(), m_difference.end(), 0.0);

    for (std::size_t tau = 1; tau <= m_maxLag; ++tau) {
        double sum = 0.0;
        for (std::size_t i = 0; i < m_bufferSize - tau; ++i) {
            const double delta = static_cast<double>(samples[i]) - static_cast<double>(samples[i + tau]);
            sum += delta * delta;
        }
        m_difference[tau] = sum;
    }

    m_difference[0] = 0.0;
}

void YinPitchDetector::computeCumulativeMeanNormalized() {
    m_cumulative[0] = 1.0;

    double runningSum = 0.0;
    for (std::size_t tau = 1; tau <= m_maxLag; ++tau) {
        runningSum += m_difference[tau];
        if (runningSum == 0.0) {
            m_cumulative[tau] = 1.0;
        } else {
            m_cumulative[tau] = m_difference[tau] * static_cast<double>(tau) / runningSum;
        }
    }
}

std::size_t YinPitchDetector::absoluteThreshold(double& probability) const {
    std::size_t candidate = 0;
    for (std::size_t tau = 2; tau < m_cumulative.size(); ++tau) {
        if (m_cumulative[tau] < m_threshold) {
            while (tau + 1 < m_cumulative.size() && m_cumulative[tau + 1] < m_cumulative[tau]) {
                ++tau;
            }
            probability = 1.0 - m_cumulative[tau];
            return tau;
        }
    }

    double minValue = std::numeric_limits<double>::infinity();
    for (std::size_t tau = 2; tau < m_cumulative.size(); ++tau) {
        if (m_cumulative[tau] < minValue) {
            minValue = m_cumulative[tau];
            candidate = tau;
        }
    }

    if (std::isfinite(minValue)) {
        probability = 1.0 - minValue;
    } else {
        probability = 0.0;
        candidate = 0;
    }

    return candidate;
}

double YinPitchDetector::parabolicInterpolation(std::size_t tau, const std::vector<double>& values) {
    if (tau == 0 || tau >= values.size() - 1) {
        return static_cast<double>(tau);
    }

    const double s0 = values[tau - 1];
    const double s1 = values[tau];
    const double s2 = values[tau + 1];

    const double denominator = (s0 + s2) - 2.0 * s1;
    if (denominator == 0.0) {
        return static_cast<double>(tau);
    }

    const double adjustment = 0.5 * (s0 - s2) / denominator;
    return static_cast<double>(tau) + adjustment;
}

double YinPitchDetector::midiFromFrequency(double frequency) {
    if (frequency <= 0.0) {
        return 0.0;
    }
    return 69.0 + 12.0 * std::log2(frequency / 440.0);
}

std::string YinPitchDetector::noteNameFromMidi(double midi) {
    if (!std::isfinite(midi)) {
        return {};
    }

    static constexpr std::array<const char*, 12> kNoteNames = {
        "C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"
    };

    const int roundedMidi = static_cast<int>(std::lround(midi));
    const int noteIndex = ((roundedMidi % 12) + 12) % 12;
    const int octave = (roundedMidi / 12) - 1;

    std::ostringstream oss;
    oss << kNoteNames[noteIndex] << octave;
    return oss.str();
}

} // namespace tine::dsp

File: src/native/dsp/YinPitchDetector.ts
/**
 * Pure TypeScript mirror of the native YIN pitch detector. This implementation
 * is used inside unit tests so we can validate the behaviour of the native
 * detector with deterministic inputs without depending on the platform bridges.
 */

export interface PitchResult {
  isValid: boolean;
  frequency: number;
  midi: number;
  cents: number;
  probability: number;
  noteName: string;
}

const MIN_THRESHOLD = 0.001;
const MAX_THRESHOLD = 0.999;

const NOTE_NAMES = ['C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#', 'A', 'A#', 'B'] as const;

function clamp(value: number, min: number, max: number): number {
  return Math.min(Math.max(value, min), max);
}

export class YinPitchDetector {
  private readonly sampleRate: number;
  private readonly bufferSize: number;
  private readonly maxLag: number;
  private threshold: number;
  private readonly difference: Float64Array;
  private readonly cumulative: Float64Array;
  private lastResult: PitchResult;

  constructor(sampleRate: number, bufferSize: number, threshold: number) {
    this.sampleRate = sampleRate;
    this.bufferSize = bufferSize;
    this.maxLag = Math.floor(bufferSize / 2);
    this.threshold = clamp(threshold, MIN_THRESHOLD, MAX_THRESHOLD);
    this.difference = new Float64Array(this.maxLag + 1);
    this.cumulative = new Float64Array(this.maxLag + 1);
    this.lastResult = {
      isValid: false,
      frequency: 0,
      midi: 0,
      cents: 0,
      probability: 0,
      noteName: '',
    };
  }

  setThreshold(threshold: number): void {
    this.threshold = clamp(threshold, MIN_THRESHOLD, MAX_THRESHOLD);
  }

  getLastResult(): PitchResult {
    return this.lastResult;
  }

  processBuffer(samples: ArrayLike<number>, numSamples: number): PitchResult {
    const emptyResult: PitchResult = {
      isValid: false,
      frequency: 0,
      midi: 0,
      cents: 0,
      probability: 0,
      noteName: '',
    };

    if (!samples || numSamples < this.bufferSize || this.maxLag < 2 || this.sampleRate <= 0) {
      this.lastResult = emptyResult;
      return this.lastResult;
    }

    this.computeDifference(samples);
    this.computeCumulativeMeanNormalized();

    let probability = 0;
    const tau = this.absoluteThreshold((value) => {
      probability = value;
    });

    if (tau === 0) {
      this.lastResult = emptyResult;
      return this.lastResult;
    }

    let refinedTau = tau;
    if (tau > 1 && tau < this.maxLag) {
      refinedTau = this.parabolicInterpolation(tau, this.cumulative);
    }

    if (refinedTau <= 0) {
      this.lastResult = emptyResult;
      return this.lastResult;
    }

    const frequency = this.sampleRate / refinedTau;
    if (!Number.isFinite(frequency) || frequency <= 0) {
      this.lastResult = emptyResult;
      return this.lastResult;
    }

    const midi = YinPitchDetector.midiFromFrequency(frequency);
    const nearestMidi = Math.round(midi);
    const cents = (midi - nearestMidi) * 100;

    const result: PitchResult = {
      isValid: probability > 0,
      frequency,
      midi,
      cents,
      probability: clamp(probability, 0, 1),
      noteName: YinPitchDetector.noteNameFromMidi(nearestMidi),
    };

    this.lastResult = result;
    return this.lastResult;
  }

  private computeDifference(samples: ArrayLike<number>): void {
    this.difference.fill(0);

    for (let tau = 1; tau <= this.maxLag; tau += 1) {
      let sum = 0;
      for (let i = 0; i < this.bufferSize - tau; i += 1) {
        const delta = Number(samples[i]) - Number(samples[i + tau]);
        sum += delta * delta;
      }
      this.difference[tau] = sum;
    }

    this.difference[0] = 0;
  }

  private computeCumulativeMeanNormalized(): void {
    this.cumulative[0] = 1;
    let runningSum = 0;

    for (let tau = 1; tau <= this.maxLag; tau += 1) {
      runningSum += this.difference[tau];
      if (runningSum === 0) {
        this.cumulative[tau] = 1;
      } else {
        this.cumulative[tau] = (this.difference[tau] * tau) / runningSum;
      }
    }
  }

  private absoluteThreshold(updateProbability: (probability: number) => void): number {
    for (let tau = 2; tau < this.cumulative.length; tau += 1) {
      if (this.cumulative[tau] < this.threshold) {
        while (
          tau + 1 < this.cumulative.length &&
          this.cumulative[tau + 1] < this.cumulative[tau]
        ) {
          tau += 1;
        }
        updateProbability(1 - this.cumulative[tau]);
        return tau;
      }
    }

    let minValue = Number.POSITIVE_INFINITY;
    let candidate = 0;

    for (let tau = 2; tau < this.cumulative.length; tau += 1) {
      if (this.cumulative[tau] < minValue) {
        minValue = this.cumulative[tau];
        candidate = tau;
      }
    }

    if (Number.isFinite(minValue)) {
      updateProbability(1 - minValue);
    } else {
      updateProbability(0);
      candidate = 0;
    }

    return candidate;
  }

  private parabolicInterpolation(tau: number, values: Float64Array): number {
    if (tau <= 0 || tau >= values.length - 1) {
      return tau;
    }

    const s0 = values[tau - 1];
    const s1 = values[tau];
    const s2 = values[tau + 1];
    const denominator = s0 + s2 - 2 * s1;

    if (denominator === 0) {
      return tau;
    }

    const adjustment = (0.5 * (s0 - s2)) / denominator;
    return tau + adjustment;
  }

  static midiFromFrequency(frequency: number): number {
    if (frequency <= 0) {
      return 0;
    }

    return 69 + 12 * Math.log2(frequency / 440);
  }

  static noteNameFromMidi(midi: number): string {
    if (!Number.isFinite(midi)) {
      return '';
    }

    const rounded = Math.round(midi);
    const noteIndex = ((rounded % 12) + 12) % 12;
    const octave = Math.floor(rounded / 12) - 1;
    return `${NOTE_NAMES[noteIndex]}${octave}`;
  }
}

export default YinPitchDetector;
